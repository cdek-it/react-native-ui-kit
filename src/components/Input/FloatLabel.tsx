import {
  memo,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from 'react'
import {
  Pressable,
  StyleSheet,
  type TextInput,
  type TextInputProps,
  type ViewStyle,
} from 'react-native'
import Animated, {
  interpolate,
  interpolateColor,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'

import { useMakeTestId } from '../../hooks/useMakeTestId'
import { makeStyles } from '../../utils/makeStyles'

import { InputTextBase } from './InputTextBase/InputTextBase'
import type { InputTextBaseProps } from './InputTextBase/types'

export interface FloatLabelProps extends InputTextBaseProps {
  /** Текст плейсхолдера */
  placeholder: string
  /** Показать индикатор загрузки в поле */
  loading?: boolean
  /** Дополнительная стилизация для контейнера компонента */
  style?: ViewStyle
}

/**
 * Компонент для ввода текста с плавающим плейсхолдером
 * @link https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=2168-9279
 * @see InputTextBaseProps
 */
export const FloatLabel = memo<FloatLabelProps>(
  ({
    style,
    onChangeText: onChangeTextProp,
    multiline,
    disabled,
    state,
    inputRef: inputRefProp,
    onBlur: onBlurProp,
    onFocus: onFocusProp,
    value: valueProp,
    placeholder,
    ...otherProps
  }) => {
    const styles = useStyles()
    const labelAnimation = useSharedValue(0)
    const labelColorAnimation = useSharedValue(0)
    const [focused, setFocused] = useState(false)
    const [valueState, setValueState] = useState(otherProps.defaultValue || '')
    const inputRef = useRef<TextInput>(null)

    const value = useMemo(
      () => valueProp ?? valueState,
      [valueProp, valueState]
    )

    useEffect(() => {
      labelAnimation.value = withTiming(focused || value ? 1 : 0, {
        duration: 100,
      })
    }, [focused, labelAnimation, value])

    useEffect(() => {
      labelColorAnimation.value = withTiming(
        focused && state !== 'danger' ? 1 : 0,
        { duration: 100 }
      )
    }, [focused, labelColorAnimation, state])

    const labelAnimatedStyle = useAnimatedStyle(() => {
      const labelBaseStyles = value
        ? styles.labelReducedAnimatedStyles
        : styles.labelAnimatedStyles

      return {
        top: interpolate(
          labelAnimation.value,
          [0, 1],
          [labelBaseStyles.top, styles.labelReducedAnimatedStyles.top]
        ),
        fontSize: interpolate(
          labelAnimation.value,
          [0, 1],
          [labelBaseStyles.fontSize, styles.labelReducedAnimatedStyles.top]
        ),
        opacity: interpolate(
          labelAnimation.value,
          [0, 1],
          [labelBaseStyles.opacity, styles.labelReducedAnimatedStyles.opacity]
        ),
        color: interpolateColor(
          labelColorAnimation.value,
          [0, 1],
          [styles.label.color, styles.labelFocused.color]
        ),
      }
    })

    const onContainerPress = useCallback(() => {
      inputRef.current?.focus()
    }, [])

    const onFocus = useCallback<NonNullable<TextInputProps['onBlur']>>(
      (event) => {
        setFocused(true)
        onFocusProp?.(event)
      },
      [onFocusProp]
    )

    const onBlur = useCallback<NonNullable<TextInputProps['onBlur']>>(
      (event) => {
        setFocused(false)
        onBlurProp?.(event)
      },
      [onBlurProp]
    )

    const onChangeText = useCallback<
      NonNullable<TextInputProps['onChangeText']>
    >(
      (nextValue) => {
        setValueState(nextValue)
        onChangeTextProp?.(nextValue)
      },
      [onChangeTextProp]
    )

    const { makeTestId } = useMakeTestId(
      otherProps.testID || FloatLabelTestId.root
    )

    useImperativeHandle(inputRefProp, () => inputRef.current, [])

    return (
      <Pressable
        accessible={false}
        disabled={disabled}
        style={style}
        testID={makeTestId(FloatLabelTestId.pressable)}
        onPress={onContainerPress}
      >
        <InputTextBase
          {...otherProps}
          accessibilityLabel={placeholder}
          containerStyle={StyleSheet.flatten([
            styles.inputContainer,
            multiline && styles.multilineInputContainer,
            otherProps.containerStyle,
          ])}
          disabled={disabled}
          inputRef={inputRef}
          inputStyle={StyleSheet.flatten([
            styles.input,
            multiline && styles.multilineInput,
          ])}
          multiline={multiline}
          placeholder=''
          state={state}
          testID={makeTestId()}
          value={value}
          onBlur={onBlur}
          onChangeText={onChangeText}
          onFocus={onFocus}
        />
        <Animated.Text
          style={[styles.label, labelAnimatedStyle]}
          testID={makeTestId(FloatLabelTestId.placeholder)}
        >
          {placeholder}
        </Animated.Text>
      </Pressable>
    )
  }
)

const useStyles = makeStyles(({ spacing, typography }) => ({
  inputContainer: { height: 56 },
  multilineInputContainer: { height: 120 },
  input: { paddingTop: 26, paddingBottom: 13 },
  multilineInput: {
    marginTop: 26,
    marginBottom: 13,
    paddingTop: 0,
    paddingBottom: 0,
  },
  label: {
    position: 'absolute',
    paddingHorizontal: spacing.Padding['p-1'],
    left: 10.5,
    color: typography.Color.Common['text-color-secondary'],
    includeFontPadding: false,
    verticalAlign: 'middle',
  },
  labelAnimatedStyles: { top: 20, fontSize: 14, opacity: 0.6 },
  labelReducedAnimatedStyles: { top: 10.5, fontSize: 12.25, opacity: 1 },
  labelFocused: { color: typography.Color.Common['text-primary'] },
}))

export const FloatLabelTestId = {
  root: 'FloatLabel',
  pressable: 'PressableContainer',
  placeholder: 'Placeholder',
}
