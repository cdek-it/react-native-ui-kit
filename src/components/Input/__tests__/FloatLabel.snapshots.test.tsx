import { fireEvent, render } from '@testing-library/react-native'

import { FloatLabel, FloatLabelTestId } from '../FloatLabel'

import useFakeTimers = jest.useFakeTimers

jest.mock('../InputTextBase/InputTextBase', () => ({
  InputTextBase: 'InputTextBase',
}))

describe('FloatLabel snapshots', () => {
  beforeEach(() => {
    useFakeTimers()
  })

  test('focused/unfocused snapshots', () => {
    const { toJSON, getByTestId } = render(
      <FloatLabel placeholder='Label' style={{ margin: 10 }} />
    )

    expect(toJSON()).toMatchSnapshot('unfocused')

    const inputTextBase = getByTestId(FloatLabelTestId.root)
    fireEvent(inputTextBase, 'focus')
    jest.runAllTimers()

    expect(toJSON()).toMatchSnapshot('focused')
  })

  test('unfocused with value snapshots, multiline, custom testID', () => {
    const { toJSON } = render(
      <FloatLabel
        multiline
        placeholder='Label'
        style={{ margin: 10 }}
        testID='CustomTestId'
        value='Value'
      />
    )
    jest.runAllTimers()

    expect(toJSON()).toMatchSnapshot('focused with value')
  })
})
