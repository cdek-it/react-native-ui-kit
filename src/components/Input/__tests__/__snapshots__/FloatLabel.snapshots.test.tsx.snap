// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatLabel snapshots focused/unfocused snapshots: focused 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={false}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "margin": 10,
    }
  }
  testID="FloatLabelPressableContainer"
>
  <InputTextBase
    accessibilityLabel="Label"
    containerStyle={
      {
        "height": 56,
      }
    }
    inputRef={
      {
        "current": null,
      }
    }
    inputStyle={
      {
        "paddingBottom": 13,
        "paddingTop": 26,
      }
    }
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder=""
    testID="FloatLabel"
    value=""
  />
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "color": "rgba(24, 135, 0, 1)",
          "fontSize": 10.5,
          "opacity": 1,
          "top": 10.5,
        },
      }
    }
    style={
      {
        "color": "rgba(0, 0, 0, 0.6)",
        "fontSize": 14,
        "includeFontPadding": false,
        "left": 10.5,
        "opacity": 0.6,
        "paddingHorizontal": 3.5,
        "position": "absolute",
        "top": 20,
        "verticalAlign": "middle",
      }
    }
    testID="FloatLabelPlaceholder"
  >
    Label
  </Text>
</View>
`;

exports[`FloatLabel snapshots focused/unfocused snapshots: unfocused 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={false}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "margin": 10,
    }
  }
  testID="FloatLabelPressableContainer"
>
  <InputTextBase
    accessibilityLabel="Label"
    containerStyle={
      {
        "height": 56,
      }
    }
    inputRef={
      {
        "current": null,
      }
    }
    inputStyle={
      {
        "paddingBottom": 13,
        "paddingTop": 26,
      }
    }
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder=""
    testID="FloatLabel"
    value=""
  />
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "color": "rgba(0, 0, 0, 0.6)",
          "fontSize": 14,
          "opacity": 0.6,
          "top": 20,
        },
      }
    }
    style={
      {
        "color": "rgba(0, 0, 0, 0.6)",
        "fontSize": 14,
        "includeFontPadding": false,
        "left": 10.5,
        "opacity": 0.6,
        "paddingHorizontal": 3.5,
        "position": "absolute",
        "top": 20,
        "verticalAlign": "middle",
      }
    }
    testID="FloatLabelPlaceholder"
  >
    Label
  </Text>
</View>
`;

exports[`FloatLabel snapshots unfocused with value snapshots, multiline, custom testID: focused with value 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={false}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "margin": 10,
    }
  }
  testID="CustomTestIdPressableContainer"
>
  <InputTextBase
    accessibilityLabel="Label"
    containerStyle={
      {
        "height": 120,
      }
    }
    inputRef={
      {
        "current": null,
      }
    }
    inputStyle={
      {
        "marginBottom": 13,
        "marginTop": 26,
        "paddingBottom": 0,
        "paddingTop": 0,
      }
    }
    multiline={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder=""
    testID="CustomTestId"
    value="Value"
  />
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "color": "rgba(0, 0, 0, 0.6)",
          "fontSize": 10.5,
          "opacity": 1,
          "top": 10.5,
        },
      }
    }
    style={
      {
        "color": "rgba(0, 0, 0, 0.6)",
        "fontSize": 12.25,
        "includeFontPadding": false,
        "left": 10.5,
        "opacity": 1,
        "paddingHorizontal": 3.5,
        "position": "absolute",
        "top": 10.5,
        "verticalAlign": "middle",
      }
    }
    testID="CustomTestIdPlaceholder"
  >
    Label
  </Text>
</View>
`;
